core/src/test/java/org/apache/iceberg/util/TestTableScanUtil.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.util;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.iceberg.BaseCombinedScanTask;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.MockFileScanTask;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.junit.Assert;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class TestTableScanUtil {
+
+  private List<FileScanTask> tasksWithDataAndDeleteSizes(List<Pair<Long, Long[]>> sizePairs) {
+    return sizePairs.stream().map(sizePair -> {
+      DataFile dataFile = dataFileWithSize(sizePair.first());
+      DeleteFile[] deleteFiles = deleteFilesWithSizes(
+          Arrays.stream(sizePair.second()).mapToLong(Long::longValue).toArray());
+      return new MockFileScanTask(dataFile, deleteFiles);
+    }).collect(Collectors.toList());
+  }
+
+  private DataFile dataFileWithSize(long size) {
+    DataFile mockFile = Mockito.mock(DataFile.class);
+    Mockito.when(mockFile.fileSizeInBytes()).thenReturn(size);
+    return mockFile;
+  }
+
+  private DeleteFile[] deleteFilesWithSizes(long... sizes) {
+    return Arrays.stream(sizes).mapToObj(size -> {
+      DeleteFile mockDeleteFile = Mockito.mock(DeleteFile.class);
+      Mockito.when(mockDeleteFile.fileSizeInBytes()).thenReturn(size);
+      return mockDeleteFile;
+    }).toArray(DeleteFile[]::new);
+  }
+
+  @Test
+  public void testPlanTaskWithDeleteFiles() {
+    List<FileScanTask> testFiles = tasksWithDataAndDeleteSizes(
+        Arrays.asList(
+            Pair.of(100L, new Long[] {20L, 20L}),
+            Pair.of(250L, new Long[0]),
+            Pair.of(100L, new Long[] {40L, 40L, 40L}),
```

Comment:
I think you want at least one file scan task that weighs over 512 with delete files alone. One with 250 data and 3 delete files of 100 each would do it. Then that shouldn't get combined with anything if weighed properly.

Reply:
@rdblue 
Sorry for late reply since I am busy these day.
I've change the unit test case as following with open cost 50, lookback 3 and target weight 300: 
```
group 1. data: 100, delete: 50, 100
group 2. data: 50, delete: 1, 50
group 3. data: 50, delete: 100
group 4. data: 1, delete: 1, 1
group 5. data: 75, delete 75
```
plan tasks after bin-packing would be `{group 1}, {group 2, group 3}, {group 4, group 5}`
1. group 1 total weight : 100 + 50 + 100 = 300
2. group 2 and group 3 total weight: `(50 + 1 + 50) -> (50 + 50 + 50)` + 50 + 100 = 300
3. group 3 and group 4 total weight: `(1 + 1 + 1) -> (50 + 50 + 50)` + 75 + 75 = 300

Reply:
It is easier to read if you don't have everything add up to exactly the target weight. That way the reader doesn't need to think about edge cases.

Reply:
This isn't a blocker, so I'll merge this when tests pass.